akka-actor/src/main/scala/akka/pattern/PromiseRef.scala

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * A combination of a Future and an ActorRef associated with it, which points
+ * to an actor performing a task which will eventually resolve the Future.
+ */
+trait FutureRef[T] {
+  /**
+   * ActorRef associated with this FutureRef.
+   */
+  def ref: ActorRef
+
+  /**
+   * Future associated with this FutureRef.
+   */
+  def future: Future[T]
+}
+
+/**
+ * A combination of a Promise and an ActorRef associated with it, which points
+ * to an actor performing a task which will eventually resolve the Promise.
+ */
+trait PromiseRef[T] { this: FutureRef[T] ⇒
+  /**
+   * ActorRef associated with this PromiseRef.
+   */
+  def ref: ActorRef
+
+  /**
+   * Promise associated with this PromiseRef.
+   */
+  def promise: Promise[T]
+
+  /**
+   * Future containing the value of the Promise associated with this PromiseRef.
+   */
+  final def future = promise.future
+
+  /**
+   * Converts this PromiseRef to FutureRef, effectively narrowing it's API.
+   */
+  def toFutureRef: FutureRef[T]
+}
+
+object PromiseRef {
+  /**
+   * Wraps an ActorRef and a Promise into a PromiseRef.
+   */
+  private[akka] def wrap[T](actorRef: ActorRef, promise: Promise[T]): PromiseRef[T] = {
+    new PromiseRefImpl(actorRef, promise)
+  }
+
+  /**
+   * Constructs a new PromiseRef which will be completed with the first message sent to it.
+   *
+   * {{{
+   * // enables transparent use of PromiseRef as ActorRef and Promise
+   * import PromiseRef.Implicits._
+   *
+   * val promiseRef = PromiseRef(system, 5.seconds)
+   * promiseRef ! "message"
+   * promiseRef.onComplete(println)  // prints "message"
+   * }}}
+   */
+  def apply(system: ActorSystem, timeout: Timeout): PromiseRef[Any] = {
+    val provider = system.asInstanceOf[ExtendedActorSystem].provider
+    AskPromiseRef(provider, timeout)
+  }
+
+  /**
+   * Constructs a new PromiseRef which will be completed with the first message sent to it.
+   *
+   * {{{
+   * // enables transparent use of PromiseRef as ActorRef and Promise
+   * import PromiseRef.Implicits._
+   *
+   * // requires an implicit ActorSystem in scope
+   * val promiseRef = PromiseRef(5.seconds)
+   * promiseRef ! "message"
+   * promiseRef.future.onComplete(println)  // prints "message"
+   * }}}
+   */
+  def apply(timeout: Timeout)(implicit system: ActorSystem): PromiseRef[Any] = {
+    PromiseRef(system, timeout)
+  }
+
+  object Implicits {
+    /**
+     * Provides a conversion to transparently use PromiseRef as Promise.
+     */
+    implicit def toPromise[T](promiseRef: PromiseRef[T]): Promise[T] = promiseRef.promise
+
+    /**
+     * Provides a conversion to transparently use PromiseRef as ActorRef.
+     */
+
+    implicit def toActorRef(promiseRef: PromiseRef[_]): ActorRef = promiseRef.ref
+
+    /**
+     * Provides a conversion to transparently use PromiseRef to dispatch messages via Scala DSL.
+     */
+    implicit def toScalaActorRef(promiseRef: PromiseRef[_]): ScalaActorRef = promiseRef.ref
+  }
```

Comment:
I’m not sure about these, we normally don’t do that. At the cost of one more import these conversions make the code one bit less obvious in order to save a few keystrokes—unless of course I’m missing something.